src/packagedcode/go_mod.py

```diff
@@ -0,0 +1,210 @@
+module is in the form
+module github.com/alecthomas/participle
+
+For example:
+>>> ob = GoMod()
+>>> p = ob.parse_module('module github.com/alecthomas/participle')
+>>> assert p.group('module') == ('github.com/alecthomas/participle')
+
+require or exclude can be in the form
+require github.com/davecgh/go-spew v1.1.1
+or
+exclude github.com/davecgh/go-spew v1.1.1
+or
+github.com/davecgh/go-spew v1.1.1
+
+For example:
+>>> ob = GoMod()
+
+>>> p = ob.parse_require('require github.com/davecgh/go-spew v1.1.1')
+>>> assert p.group('namespace') == ('github.com/davecgh')
+>>> assert p.group('name') == ('go-spew')
+>>> assert p.group('version') == ('v1.1.1')
+
+>>> p = ob.parse_exclude('exclude github.com/davecgh/go-spew v1.1.1')
+>>> assert p.group('namespace') == ('github.com/davecgh')
+>>> assert p.group('name') == ('go-spew')
+>>> assert p.group('version') == ('v1.1.1')
+
+>>> p = ob.parse_dep_link('github.com/davecgh/go-spew v1.1.1')
+>>> assert p.group('namespace') == ('github.com/davecgh')
+>>> assert p.group('name') == ('go-spew')
+>>> assert p.group('version') == ('v1.1.1')
+"""
+
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class GoMod(object):
+    # Regex expressions to parse different types of dependency
+    parse_module = re.compile(
+        r'^module\s'
+        r'(?P<module>[a-z].*)'
+    ).match
+
+    parse_module_name = re.compile(
+        r'^module(\s)*'
+        r'(?P<namespace>(.*))'
+        r'/'
+        r'(?P<name>[^\s]*)'
+    ).match
+
+    parse_require = re.compile(
+        r'^require(\s)*'
+        r'(?P<namespace>(.*))'
+        r'/'
+        r'(?P<name>[^\s]*)'
+        r'\s'
+        r'(?P<version>(.*))'
+    ).match
+
+    parse_exclude = re.compile(
+        r'^exclude(\s)*'
+        r'(?P<namespace>(.*))'
+        r'/'
+        r'(?P<name>[^\s]*)'
+        r'\s'
+        r'(?P<version>(.*))'
+    ).match
+
+    parse_dep_link = re.compile(
+        r'(?P<namespace>(.*))'
+        r'/'
+        r'(?P<name>[^\s]*)'
+        r'\s'
+        r'(?P<version>(.*))'
+    ).match
+
+    def preprocess(self, line):
+        """
+        Return line string after removing commented portion and excess spaces.
+        """
+        if "//" in line:
+            line = line[:line.index('//')]
+        line = line.strip()
+        return line
+
+    def parse_gomod(self, location):
+        """
+        Return a dictionary containing all the important go.mod file data.
+        """
+        with io.open(location, encoding='utf-8', closefd=True) as data:
+            lines = data.readlines()
+
+        gomod_data = {}
+        require = []
+        exclude = []
+
+        for i, line in enumerate(lines):
+            line = self.preprocess(line)
+            parsed_module = self.parse_module(line)
+            if parsed_module:
+                gomod_data['module'] = parsed_module.group('module')
```

Comment:
What does `.group()` do?

Reply:
@majurg 
`.group()` method returns a tuple containing all the subgroups of the match, from 1 up to however many groups are in the pattern. The default argument is used for groups that did not participate in the match; it defaults to None. 

```
>>> m = re.match(r"(\d+)\.(\d+)", "27.1835")
>>> m.groups()
('27', '1835')
```